--- a/res/res_musiconhold.c
+++ b/res/res_musiconhold.c
@@ -1369,26 +1369,26 @@ static int local_ast_moh_start(struct ast_channel *chan, const char *mclass, con
 	 * 4) The default class.
 	 */
 	if (!ast_strlen_zero(ast_channel_musicclass(chan))) {
-		mohclass = get_mohbyname(ast_channel_musicclass(chan), 1, 0);
+		mohclass = get_mohbyname(ast_channel_musicclass(chan), !realtime_possible, 0);
 		if (!mohclass && realtime_possible) {
 			var = ast_load_realtime("musiconhold", "name", ast_channel_musicclass(chan), SENTINEL);
 		}
 	}
 	if (!mohclass && !var && !ast_strlen_zero(mclass)) {
-		mohclass = get_mohbyname(mclass, 1, 0);
+		mohclass = get_mohbyname(mclass, !realtime_possible, 0);
 		if (!mohclass && realtime_possible) {
 			var = ast_load_realtime("musiconhold", "name", mclass, SENTINEL);
 		}
 	}
 	if (!mohclass && !var && !ast_strlen_zero(interpclass)) {
-		mohclass = get_mohbyname(interpclass, 1, 0);
+		mohclass = get_mohbyname(interpclass, !realtime_possible, 0);
 		if (!mohclass && realtime_possible) {
 			var = ast_load_realtime("musiconhold", "name", interpclass, SENTINEL);
 		}
 	}
 
 	if (!mohclass && !var) {
-		mohclass = get_mohbyname("default", 1, 0);
+		mohclass = get_mohbyname("default", !realtime_possible, 0);
 		if (!mohclass && realtime_possible) {
 			var = ast_load_realtime("musiconhold", "name", "default", SENTINEL);
 		}
