--- a/res/res_pjsip_endpoint_identifier_user.c
+++ b/res/res_pjsip_endpoint_identifier_user.c
@@ -69,6 +69,11 @@ static struct ast_sip_endpoint *username_identify(pjsip_rx_data *rdata)
 		return NULL;
 	}
 
+    /* Just search the endpoint name */
+    if ((endpoint = ast_sorcery_retrieve_by_id(ast_sip_get_sorcery(), "endpoint", endpoint_name))) {
+        goto done;
+    }
+
 	/* Attempt to find the endpoint given the name and domain provided */
 	snprintf(id, sizeof(id), "%s@%s", endpoint_name, domain_name);
 	if ((endpoint = ast_sorcery_retrieve_by_id(ast_sip_get_sorcery(), "endpoint", id))) {
